src/main/scala/kamon/prometheus/PrometheusPushgatewayReporter.scala

```diff
@@ -0,0 +1,58 @@
+package kamon.prometheus
+
+import com.typesafe.config.Config
+import kamon.Kamon
+import kamon.metric.PeriodSnapshot
+import kamon.module.{MetricReporter, Module, ModuleFactory}
+import org.slf4j.LoggerFactory
+
+class PrometheusPushgatewayReporter(
+    configPath: String,
+    pushgatewayPath: String,
+    @volatile private var httpClientFactory: Config => HttpClient
+  ) extends MetricReporter {
+
+  private val _logger = LoggerFactory.getLogger(classOf[PrometheusPushgatewayReporter])
+
+  @volatile private var httpClient: HttpClient = _
+  @volatile private var settings: PrometheusSettings.Generic = _
+
+  {
+    reconfigure(Kamon.config())
+  }
+
+  def this(httpClientFactory: Config => HttpClient) =
+    this("kamon.prometheus", "pushgateway", httpClientFactory)
+
+  override def reportPeriodSnapshot(snapshot: PeriodSnapshot): Unit = {
+    val scrapeDataBuilder = new ScrapeDataBuilder(settings, PrometheusSettings.environmentTags(settings))
```

Comment:
Hey @peavers, I was taking a look at the documentation on the pushgateway and it seemed like the data being pushed to the PG has to be the same as what would have been exposed in the scrape endpoint but with this implementation we would be overwriting the data every minute. For example, if a job runs for 5 minutes, each minute we would be pushing what happened in the last minute and that would overwrite the data from the previous minute (at least for the metrics that have already been reported).. is that the intended flow? maybe I misunderstood the docs!

Reply:
Hi @ivantopo 

We've taken a look at a few other implementations and they seem to be doing the same thing when it comes to collecting and pushing the metrics: 

Java: https://github.com/prometheus/client_java/blob/fbc3fc42f8470b55b5274c9921f6202f17ffc1bc/simpleclient_pushgateway/src/main/java/io/prometheus/client/exporter/PushGateway.java#L192

C#: 
https://github.com/prometheus-net/prometheus-net/blob/505a08e96878d2811afcd02e2fa01efd867fddf1/Prometheus.NetStandard/MetricPusher.cs#L67

So seems to be the normal way of handling it? Would love to know what you think about this, or if you've got an alternative solution/suggestion we could try out? 

Cheers

Reply:
Hey @peavers! I think that the pushing part is fine, the only part that doesn't seem to fit correctly is the "when". Let me explain:
  - From the Java example (see [this section](https://github.com/kamon-io/Kamon/blob/master/kamon-core/src/main/scala/kamon/Init.scala#L33-L36) it looks like the intended way to use it is to push the metrics only once after the entire batch job has finished. If this is the desired behavior then `reportPeriodSnapshot` should accumulate all the data it receives and push it out only when `stop` is called.
  - From the .Net example, it seems like all data is being [pushed periodically](https://github.com/prometheus-net/prometheus-net/blob/505a08e96878d2811afcd02e2fa01efd867fddf1/Prometheus.NetStandard/MetricPusher.cs#L80), which looks more similar to what is happening on this PR, just with the exception here we are not accumulating the data.

The part that I find missing is [this one](https://github.com/kamon-io/kamon-prometheus/blob/master/src/main/scala/kamon/prometheus/PrometheusReporter.scala#L63-L66), which is the one that ensures that data from the entire run is going to be sent instead of only the data from the last snapshot (in Kamon core we only report the data that was recorded within the last interval instead of the cumulative values from the start of the app, that's why we need to accumulate it for Prometheus). Hope this explains it better! We could also jump on a quick chat to hash this out and finally get this merged!